crates/red_knot_python_semantic/src/types/call/bind.rs

```diff
@@ -183,23 +482,16 @@ pub(crate) struct CallableBinding<'db> {
     overloads: SmallVec<[Binding<'db>; 1]>,
 }
 
-impl<'db> CallableBinding<'db> {
-    /// Bind a [`CallArguments`] against a [`CallableSignature`].
-    ///
-    /// The returned [`CallableBinding`] provides the return type of the call, the bound types for
-    /// all parameters, and any errors resulting from binding the call.
-    fn bind(
-        db: &'db dyn Db,
+impl<'call, 'db> CallableBinding<'call, 'db> {
+    fn match_parameters(
         signature: &CallableSignature<'db>,
-        arguments: &CallArguments<'_, 'db>,
+        mut arguments: CallArguments<'call, 'db>,
     ) -> Self {
         // If this callable is a bound method, prepend the self instance onto the arguments list
         // before checking.
-        let arguments = if let Some(bound_type) = signature.bound_type {
-            Cow::Owned(arguments.with_self(bound_type))
-        } else {
-            Cow::Borrowed(arguments)
-        };
+        if let Some(bound_type) = signature.bound_type {
+            arguments = arguments.with_self(bound_type);
```

Comment:
It seems that this path might now panic if `arguments` already has `self` bound. 

We should at least document this constraint on `match_parameters` but it makes me think if we should change the signature to return a `Result` or `Option` here to have the caller *guarantee* that `arguments` isn't bound yet.

Reply:
This can no longer panic, and we no longer clone.  Instead we `push_front` onto the `VecDeque` and `pop_front` before returning